### PR TITLE
Smart layout API

### LINKED ISSUE #640: Smart layout API

### PR DESCRIPTION
**Pull request by jedrichards**
Closes #640

### CODE REVIEW

**Comment by bohemian-coding-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>This pull request may need a CHANGELOG entry.</td>
    </tr>
  
<tr>
      <td>:warning:</td>
      <td>

  `Source/dom/layers/SymbolMaster.js` changed, but not:
  - 📚 its docs
- 🧪 its tests

That's OK as long as you're refactoring.
  </td>
    </tr>
  
<tr>
      <td>:warning:</td>
      <td>

  `Source/dom/models/SmartLayout.js` changed, but not:
  - 📚 its docs
- 🧪 its tests

That's OK as long as you're refactoring.
  </td>
    </tr>
  </tbody>
</table>



<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 351dadf0c7dbf8c345f049c3bc401a52da6d9bcc
</p>

File: docs/api/Group.md [hunk 1]
```diff
@@ -10,24 +10,25 @@ var Group = require('sketch/dom').Group
...
-| transform.flippedVertically<span class="arg-type">boolean</span>            | If the Group is vertically flipped.                                                             |
-| style<span class="arg-type">[Style](#style)</span>                          | The style of the Group.                                                                         |
-| sharedStyleId<span class="arg-type">string / null</span>                    | The ID of the [SharedStyle](#shared-style) this Group is linked to if any.                      |
-| layers<span class="arg-type">[Layer](#layer)[]</span>                       | The layers that this component groups together.                                                 |
+| Properties |  |
+| --- | --- |
+| id<span class="arg-type">string</span> | The unique ID of the Group. |
+| name<span class="arg-type">string</span> | The name of the Group |
+| parent<span class="arg-type">[Group](#group)</span> | The group the Group is in. |
+| locked<span class="arg-type">boolean</span> | If the group is locked. |
```
Comment (christianklotz) [inline docs/api/Group.md]:
```suggestion
| locked<span class="arg-type">boolean</span> | If the Group is locked. |
```

File: docs/api/Group.md [hunk 2]
```diff
@@ -10,24 +10,25 @@ var Group = require('sketch/dom').Group
...
+| hidden<span class="arg-type">boolean</span> | If the group is hidden. |
```
Comment (christianklotz) [inline docs/api/Group.md]:
```suggestion
| hidden<span class="arg-type">boolean</span> | If the Group is hidden. |
```

File: docs/api/Group.md [hunk 3]
```diff
@@ -10,24 +10,25 @@ var Group = require('sketch/dom').Group
...
+| frame<span class="arg-type">[Rectangle](#rectangle)</span> | The frame of the Group. This is given in coordinates that are local to the parent of the layer. |
```
Comment (christianklotz) [inline docs/api/Group.md]:
```suggestion
| frame<span class="arg-type">[Rectangle](#rectangle)</span> | The frame of the Group. This is given in coordinates that are local to the parent of the Layer. |
```

File: docs/api/Group.md [hunk 4]
```diff
@@ -10,24 +10,25 @@ var Group = require('sketch/dom').Group
...
+| selected<span class="arg-type">boolean</span> | If the Group is selected. |
+| flow<span class="arg-type">[Flow](#flow)</span> | The prototyping action associated with the Group. |
+| exportFormats<span class="arg-type">[ExportFormat](#export-format)[]</span> | The export formats of the Group. |
+| transform<span class="arg-type">object</span> | The transformation applied to the Group. |
+| transform.rotation<span class="arg-type">number</span> | The rotation of the Group in degrees, clock-wise. |
+| transform.flippedHorizontally<span class="arg-type">boolean</span> | If the Group is horizontally flipped. |
+| transform.flippedVertically<span class="arg-type">boolean</span> | If the Group is vertically flipped. |
+| style<span class="arg-type">[Style](#style)</span> | The style of the Group. |
+| sharedStyleId<span class="arg-type">string / null</span> | The ID of the [SharedStyle](#shared-style) this Group is linked to if any. |
+| layers<span class="arg-type">[Layer](#layer)[]</span> | The layers that this component groups together. |
```
Comment (christianklotz) [inline docs/api/Group.md]:
```suggestion
| layers<span class="arg-type">[Layer](#layer)[]</span> | The Layers that this component groups together. |
```

File: docs/api/Group.md [hunk 5]
```diff
@@ -10,24 +10,25 @@ var Group = require('sketch/dom').Group
...
+| smartLayout<span class="arg-type">[SmartLayout](#smartlayout)</span> | The Group's smart layout. |
```
Comment (christianklotz) [inline docs/api/Group.md]:
```suggestion
| smartLayout<span class="arg-type">[SmartLayout](#smartlayout)</span> | The Group's Smart Layout. |
```

File: docs/api/SmartLayout.md [hunk 6]
```diff
@@ -0,0 +1,43 @@
+---
+title: SmartLayout
+order: 215
+section: models
+---
+
+The [SmartLayout](#smartlayout) object contains the set of possible smart layouts that can be applied to [SymbolMaster](#symbol-master) and [Group](#group) layers.
```
Comment (christianklotz) [inline docs/api/SmartLayout.md]:
```suggestion
The [SmartLayout](#smartlayout) object contains the set of possible Smart Layouts that can be applied to [SymbolMaster](#symbol-master) and [Group](#group) layers.
```

File: docs/api/SmartLayout.md [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+| BottomToTop        | Smart layout flowing from bottom to top             |
+
+## Set a smart layout
+
+```javascript
+const SmartLayout = require('sketch').SmartLayout
+layer.smartLayout = SmartLayout.TopToBottom
+```
+
+Given a reference to a [SymbolMaster](#symbol-master) or [Group](#group) layer use the `smartLayout` setter to apply one of the smart layout values.
```
Comment (christianklotz) [inline docs/api/SmartLayout.md]:
```suggestion
Given a reference to a [SymbolMaster](#symbol-master) or [Group](#group) layer use the `smartLayout` setter to apply one of the Smart Layout values.
```

File: docs/api/SmartLayout.md [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+## Clear a smart layout
```
Comment (christianklotz) [inline docs/api/SmartLayout.md]:
```suggestion
## Clear a Smart Layout
```

File: docs/api/SmartLayout.md [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+| Properties         |                                                     |
+| ------------------ | --------------------------------------------------- |
+| LeftToRight        | Smart layout flowing left to right                  |
+| HorizontallyCenter | Smart layout expanding horizontally from the center |
+| RightToLeft        | Smart layout flowing right to left                  |
+| TopToBottom        | Smart layout flowing from top to bottom             |
+| VerticallyCenter   | Smart layout expanding verically from the center    |
+| BottomToTop        | Smart layout flowing from bottom to top             |
+
+## Set a smart layout
```
Comment (christianklotz) [inline docs/api/SmartLayout.md]:
```suggestion
## Set a Smart Layout
```

File: docs/api/SmartLayout.md [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+The [SmartLayout](#smartlayout) object contains the set of possible smart layouts that can be applied to [SymbolMaster](#symbol-master) and [Group](#group) layers.
+
+| Properties         |                                                     |
+| ------------------ | --------------------------------------------------- |
+| LeftToRight        | Smart layout flowing left to right                  |
+| HorizontallyCenter | Smart layout expanding horizontally from the center |
+| RightToLeft        | Smart layout flowing right to left                  |
+| TopToBottom        | Smart layout flowing from top to bottom             |
+| VerticallyCenter   | Smart layout expanding verically from the center    |
+| BottomToTop        | Smart layout flowing from bottom to top             |
```
Comment (christianklotz) [inline docs/api/SmartLayout.md]: Replace with _Smart Layout_ for consistency.

File: docs/api/SmartLayout.md [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+```javascript
+const SmartLayout = require('sketch').SmartLayout
+layer.smartLayout = SmartLayout.TopToBottom
+```
+
+Given a reference to a [SymbolMaster](#symbol-master) or [Group](#group) layer use the `smartLayout` setter to apply one of the smart layout values.
+
+## Clear a smart layout
+
+Set the `smartLayout` value to `null` to remove the smart layout. This is the equivalent of selecting "None" in the UI.
```
Comment (christianklotz) [inline docs/api/SmartLayout.md]:
```suggestion
Set the `smartLayout` value to `null` to remove the Smart Layout. This is the equivalent of selecting "None" in the Sketch Inspector.
```

Maybe also worth clarifying that the dimensions remain the same after clearing the layout?

File: docs/api/SmartLayout.md [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+
+## Trigger a smart layout
+
+```javascript
+const SmartLayout = require('sketch').SmartLayout
+symbolMaster.smartLayout = SmartLayout.TopToBottom
+symbolInstance.resizeWithSmartLayout()
+```
+
+In order to trigger a smart layout resize in an instance, for example after changing an override value, call the `resizeWithSmartLayout()` method on the [SymbolInstance](#symbol-instance) layer.
```
Comment (christianklotz) [inline docs/api/SmartLayout.md]:
```suggestion
In order to trigger a Smart Layout resize in an instance, for example after changing an override value, call the `resizeWithSmartLayout()` method on the [SymbolInstance](#symbol-instance) layer.
```

File: docs/api/SymbolInstance.md [hunk 13]
```diff
@@ -80,11 +80,19 @@ instance.setOverrideValue(instance.overrides[0], 'overridden')
...
+| Parameters |  |
+| --- | --- |
+| override<span class="arg-type">[Override](#symbol-override) - required</span> | The override to change. |
+| value<span class="arg-type">string / NSImage - required</span> | The value of override to set. Can be a string or an NSImage or a symbolId depending on the type of the override. |
 
 ### Returns
 
 The current Symbol instance (useful if you want to chain the calls).
+
+## Trigger a smart layout
```
Comment (christianklotz) [inline docs/api/SymbolInstance.md]:
```suggestion
## Trigger a Smart Layout
```

File: docs/api/SymbolInstance.md [hunk 14]
```diff
@@ -80,11 +80,19 @@ instance.setOverrideValue(instance.overrides[0], 'overridden')
...
+
+```javascript
+instance.resizeWithSmartLayout()
+```
+
+In order to trigger a smart layout resize in an instance, for example after changing an override value, call the `resizeWithSmartLayout()` method.
```
Comment (christianklotz) [inline docs/api/SymbolInstance.md]:
```suggestion
In order to trigger a Smart Layout resize in an instance, for example after changing an override value, call the `resizeWithSmartLayout()` method.
```

File: Source/dom/layers/Group.js [hunk 15]
```diff
@@ -101,3 +104,46 @@ Group.define('layers', {
     return item
   },
 })
+
+Group.define('smartLayout', {
```
Comment (mathieudutour) [inline Source/dom/layers/Group.js]: I think you need to remove that property from Artboard and Page (which inherit from Group IIRC)

Comment (jedrichards) [reply Source/dom/layers/Group.js]: Ah gotcha, thanks

**Pull request merged**